### PR TITLE
[FLINK-5219][TableAPI&SQL] Add non-grouped session windows for batch …

### PR DESCRIPTION
**Pull request by sunjincheng121**
…tables

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [×] General
  - The pull request references the related JIRA issue ("[FLINK-5219][TableAPI&SQL] Add non-grouped session windows for batch tables")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [×] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by sunjincheng121** [comment]
@fhueske @twalthr  In this PR，I In order to code reuse "DataSetSessionWindowAggregatePreProcessor" has implemented  the "MapPartitionFunction" and "GroupCombineFunction" two interfaces, I'm not sure if it's correct， what do you think?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 1]
```diff
@@ -291,13 +292,13 @@ object AggregateUtil {
...
 
-        new DataSetSessionWindowAggregateCombineGroupFunction(
+        new DataSetSessionWindowAggregatePreProcessor(
           aggregates,
           groupings,
           // the addition two fields are used to store window-start and window-end attributes
           intermediateRowArity + 2,
           asLong(gap),
-          combineReturnType)
+          combineReturnType).asInstanceOf[GroupCombineFunction[Row,Row]]
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: Can we leave out the `asInstanceOf` ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 2]
```diff
@@ -306,6 +307,85 @@ object AggregateUtil {
...
+              inputType,
+              Option(Array(BasicTypeInfo.LONG_TYPE_INFO, BasicTypeInfo.LONG_TYPE_INFO)))
+
+          new DataSetSessionWindowAggregatePreProcessor(
+            aggregates,
+            Array(),
+            // the addition two fields are used to store window-start and window-end attributes
+            intermediateRowArity + 2,
+            asLong(gap),
+            preMapReturnType).asInstanceOf[MapPartitionFunction[Row, Row]]
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: Can we leave out the `asInstanceOf` ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 3]
```diff
@@ -306,6 +307,85 @@ object AggregateUtil {
...
+            aggregates,
+            Array(),
+            aggOffsetMapping,
+            // the additional two fields are used to store window-start and window-end attributes
+            intermediateRowArity + 2,
+            outputType.getFieldCount,
+            startPos,
+            endPos,
+            asLong(gap),
+            isInputCombined).asInstanceOf[MapPartitionFunction[Row, Row]]
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: Can we leave out the `asInstanceOf` ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregatePreProcessor.scala [hunk 4]
```diff
@@ -19,30 +19,36 @@ package org.apache.flink.table.runtime.aggregate
...
-class DataSetSessionWindowAggregateCombineGroupFunction(
+class DataSetSessionWindowAggregatePreProcessor(
     aggregates: Array[Aggregate[_ <: Any]],
     groupingKeys: Array[Int],
     intermediateRowArity: Int,
     gap: Long,
     @transient intermediateRowType: TypeInformation[Row])
-  extends RichGroupCombineFunction[Row,Row] with ResultTypeQueryable[Row] {
+  extends AbstractRichFunction
+    with MapPartitionFunction[Row,Row]
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregatePreProcessor.scala]: I would like to indent `with` and `extends` on the same indent.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateProcessor.scala [hunk 5]
```diff
@@ -59,7 +61,9 @@ class DataSetSessionWindowAggregateReduceGroupFunction(
     finalRowWindowEndPos: Option[Int],
     gap:Long,
     isInputCombined: Boolean)
-  extends RichGroupReduceFunction[Row, Row] {
+  extends AbstractRichFunction
+    with MapPartitionFunction[Row, Row]
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateProcessor.scala]: I would like to indent `with` and `extends` on the same indent.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/dataset/DataSetWindowAggregateITCase.scala [hunk 6]
```diff
@@ -141,16 +141,18 @@ class DataSetWindowAggregateITCase(configMode: TableConfigMode)
...
     val env = ExecutionEnvironment.getExecutionEnvironment
     val tEnv = TableEnvironment.getTableEnvironment(env, config)
     val table = env.fromCollection(data).toTable(tEnv, 'long, 'int, 'string)
     val windowedTable =table
       .window(Session withGap 7.milli on 'long as 'w)
       .groupBy('w)
-      .select('string.count).toDataSet[Row].collect()
+      .select('string.count)
+    val results = windowedTable.toDataSet[Row].collect()
+    val expected = "6\n1";
```
Comment (wuchong) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/dataset/DataSetWindowAggregateITCase.scala]: remove the semicolon `;`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 7]
```diff
@@ -210,14 +210,14 @@ object AggregateUtil {
...
-            intermediateRowArity - 1,
+            intermediateRowArity,
             asLong(size),
             startPos,
             endPos,
             aggregates,
             groupingOffsetMapping,
             aggOffsetMapping,
-            intermediateRowArity,
+            intermediateRowArity + 1, // the additional field is used to store the time attribute
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: good catch !

**Comment by sunjincheng121** [comment]
@wuchong thanks for review the PR. your comments make sense for me. I had update the PR.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateProcessor.scala [hunk 8]
```diff
@@ -19,15 +19,17 @@ package org.apache.flink.table.runtime.aggregate
 
 import java.lang.Iterable
 
-import org.apache.flink.api.common.functions.RichGroupReduceFunction
+import org.apache.flink.api.common.functions.{AbstractRichFunction, GroupReduceFunction,
+MapPartitionFunction}
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateProcessor.scala]: do not break the line

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/dataset/DataSetWindowAggregateITCase.scala [hunk 9]
```diff
@@ -141,16 +141,18 @@ class DataSetWindowAggregateITCase(configMode: TableConfigMode)
...
     val env = ExecutionEnvironment.getExecutionEnvironment
     val tEnv = TableEnvironment.getTableEnvironment(env, config)
     val table = env.fromCollection(data).toTable(tEnv, 'long, 'int, 'string)
     val windowedTable =table
       .window(Session withGap 7.milli on 'long as 'w)
       .groupBy('w)
-      .select('string.count).toDataSet[Row].collect()
+      .select('string.count)
+    val results = windowedTable.toDataSet[Row].collect()
+    val expected = "6\n1"
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/dataset/DataSetWindowAggregateITCase.scala]: Would be good to change the query in a way that it checks that multiple sessions are correctly computed.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregatePreProcessor.scala [hunk 10]
```diff
@@ -19,30 +19,36 @@ package org.apache.flink.table.runtime.aggregate
 
 import java.lang.Iterable
 
-import org.apache.flink.api.common.functions.RichGroupCombineFunction
+import org.apache.flink.api.common.functions.{AbstractRichFunction, GroupCombineFunction,
+MapPartitionFunction}
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregatePreProcessor.scala]: No line break. Imports may exceed the 100 char limit.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 11]
```diff
@@ -294,18 +319,29 @@ class DataSetWindowAggregate(
...
+        val mapPartitionFunction = createDataSetWindowAggregationMapPartitionFunction(
+          window,
+          namedAggregates,
+          inputType,
+          rowRelDataType,
+          namedProperties,
+          isPreMapPartition = false)
+
+        mappedInput.sortPartition(rowTimeFieldPos, Order.ASCENDING).setParallelism(1)
+          .mapPartition(mapPartitionFunction).setParallelism(1)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala]: I think we can also use `.reduceGroup()` and a `GroupReduceFunction` here. Without `groupBy`, the `GroupReduceFunction` will be executed with parallelism 1.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 12]
```diff
@@ -281,9 +279,36 @@ class DataSetWindowAggregate(
...
+          rowRelDataType,
+          namedProperties,
+          isPreMapPartition = false,
+          isInputCombined = true)
+
+        mappedInput.sortPartition(rowTimeFieldPos, Order.ASCENDING)
+          .mapPartition(preMapPartitionFunction)
+          .sortPartition(windowStartPos, Order.ASCENDING).setParallelism(1)
+          .sortPartition(windowEndPos, Order.ASCENDING).setParallelism(1)
+          .mapPartition(mapPartitionFunction).setParallelism(1)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala]: I think we can also use `.reduceGroup()` and a `GroupReduceFunction` here. Without `groupBy`, the `GroupReduceFunction` will be executed with parallelism 1.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateProcessor.scala [hunk 13]
```diff
@@ -59,7 +61,9 @@ class DataSetSessionWindowAggregateReduceGroupFunction(
     finalRowWindowEndPos: Option[Int],
     gap:Long,
     isInputCombined: Boolean)
-  extends RichGroupReduceFunction[Row, Row] {
+  extends AbstractRichFunction
+  with MapPartitionFunction[Row, Row]
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateProcessor.scala]: I like the idea of implementing a joint `MapPartition` and `GroupReduce` function. However, I think it is not necessary for the final aggregation. We can also call `DataSet.reduceGroup()` which will do the same as `DataSet.mapPartition().parallelism(1)`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 14]
```diff
@@ -306,6 +307,85 @@ object AggregateUtil {
...
+            createAggregateBufferDataType(
+              Array(),
+              aggregates,
+              inputType,
+              Option(Array(BasicTypeInfo.LONG_TYPE_INFO, BasicTypeInfo.LONG_TYPE_INFO)))
+
+          new DataSetSessionWindowAggregatePreProcessor(
+            aggregates,
+            Array(),
+            // the addition two fields are used to store window-start and window-end attributes
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: the addition**al** two fields

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 15]
```diff
@@ -306,6 +307,85 @@ object AggregateUtil {
...
+            intermediateRowArity + 2,
+            asLong(gap),
+            preMapReturnType)
+
+        } else {
+          val (startPos, endPos) = computeWindowStartEndPropertyPos(properties)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: We do not need this case if we compute the final aggregates with a `GroupReduceFunction`.

**Comment by sunjincheng121** [comment]
@fhueske thanks for the review. you are right. I need rebase the code and update the PR according your comments.

**Comment by fhueske** [comment]
Hi @sunjincheng121, any news for this PR?
Thanks, Fabian

**Comment by sunjincheng121** [comment]
Hi @fhueske I'm sorry for the late reply, and thanks a lot for your reminding. I have rebased the code and update the PR according your comments. 
Best,
SunJIncheng

**Comment by fhueske** [comment]
Thanks for the update @sunjincheng121.
IMO the PR is good to merge. 

I might wait until PR #3423 has been merged which touches all aggregation functions. 
So we might need to rebase the PR another time, but I think this should not be too much work.

Thanks, Fabian

**Comment by sunjincheng121** [comment]
@fhueske Thanks for your quick review, and sounds good.

**Comment by sunjincheng121** [comment]
Hi, @fhueske I had rebase the code on PR [#3423|https://github.com/apache/flink/pull/3423]'s commit.

**Comment by fhueske** [comment]
Thanks for the update @sunjincheng121.
Merging

**Pull request closed without merging**